### ISSUE TITLE
Compare instructions parsing problems

### ISSUE THREAD

**Issue by jrkoenig** [StanfordPL/stoke#911]
The disassembler gives errors on the following instructions:

```
WARNING: Could not encode 'cmpltsd 0xb201(%rip),%xmm1        ' within 9 bytes.
WARNING: Could not encode 'cmpnless %xmm0,%xmm1' within 5 bytes.
WARNING: Could not encode 'cmpnltsd %xmm3,%xmm1' within 5 bytes.
WARNING: Could not encode 'cmpnltsd 0x229f0(%rip),%xmm1        ' within 9 bytes.
WARNING: Could not encode 'cmpnltss %xmm0,%xmm1' within 5 bytes.
WARNING: Could not encode 'cmpnltss %xmm15,%xmm4' within 6 bytes.
WARNING: Could not encode 'cmpnltss %xmm2,%xmm1' within 5 bytes.
WARNING: Could not encode 'cmpnltss %xmm9,%xmm4' within 6 bytes.
WARNING: Could not encode 'cmpnltss 0x1456d(%rip),%xmm0        ' within 9 bytes.
WARNING: Could not encode 'cmpnltss 0x15b49(%rip),%xmm1        ' within 9 bytes.
WARNING: Could not encode 'cmpnltss 0x239d4(%rip),%xmm7        ' within 9 bytes.
```

First off, src/disassembler/disassembler.cc has the following:

```
if (is_prefix(line, "cmp", 4) || is_prefix(line, "vcmp", 4)) {
    ll = regex_replace(ll, regex("(v?cmp)eq([^ ]+)"),       "$1$2 \\$0x00,$3");
    ll = regex_replace(ll, regex("(v?cmp)lt([^ ]+)"),       "$1$2 \\$0x01,$3");
    ll = regex_replace(ll, regex("(v?cmp)le([^ ]+)"),       "$1$2 \\$0x02,$3");
    ...
```

Fixing the typo in "cmp", 4 to "cmp", 3 causes other problems. This generates errors like:

```
WARNING: Could not encode 'cmpss \cmpnltssx05, %xmm9,%xmm4' within 6 bytes.
```

According to Intel manual Vol. 2A page 3-163, `CMPNLTSS xmm1, xmm2` is a psuedo op, which gets encoded as `CMPSD xmm1,xmm2, 5`, with 5 being a constant for the to NLT comparison. It seems at some point x64asm is expected to turn `cmpss \cmpnltssx05, ...` into the actual disassembly, but this doesn't happen.

In att.table in x64asm, there is are entries for cmpss, cmpsd, etc, that take the `xmm1,xmm2, imm8` argument form. There are no entries for cmpnltss, etc.

The strange codes match those found in https://svn.jacekowski.org/chromium/trunk/third_party/yasm/source/patched-yasm/modules/arch/x86/tests/avxcc.asm for translating the psudeo-ops, but the replacements are all wrong.

**Comment by bchurchill** [comment]
This probably overlaps with bugs in x64asm.  It would be a good idea to make sure that x64asm _can_ assemble the right instruction.  You can test this with the bin/asm tool in the x64asm repo.  If it can't, that's the first thing to fix.  Once that works, it's a matter of fixing the regexes in the "disassembler" so that it can turn the objdump'd code into something our assembler can parse.

**Comment by stefanheule** [comment]
There are two problems here.  One is the issue @jrkoenig identified where `"cmp", 4` should be `"cmp", 3`.  The other problem is that the regular expression is all wrong.

    ll = regex_replace(ll, regex("(v?cmp)eq([^ ]+)"),       "$1$2 \\$0x00,$3");

should be

    ll = regex_replace(ll, regex("(v?cmp)eq([^ ]+)"),       "$1$2 $$0x00");

I'll be fixing these over in #974.  The rest of the replacement looks right and seems to be working after my fixes.

**Comment by stefanheule** [comment]
And there's one more issue, the order of replacements matters.  You cannot do it this way:

    ll = regex_replace(ll, regex("(v?cmp)eq([^ ]+)"),       "$1$2 $$0x00,");
    ll = regex_replace(ll, regex("(v?cmp)eq_os([^ ]+)"),    "$1$2 $$0x10,");

Instead, the order has to be reversed.

**Comment by stefanheule** [comment]
All of these problems were fixed in #974.
